File: src/fr/adrienbrault/idea/symfony2plugin/ui/MethodParameterReferenceSettingsForm.java
package fr.adrienbrault.idea.symfony2plugin.ui;

import com.intellij.openapi.options.Configurable;
import com.intellij.openapi.options.ConfigurationException;
import com.intellij.openapi.project.Project;
import com.intellij.ui.AnActionButton;
import com.intellij.ui.AnActionButtonRunnable;
import com.intellij.ui.ToolbarDecorator;
import com.intellij.ui.table.TableView;
import com.intellij.util.ui.ColumnInfo;
import com.intellij.util.ui.ElementProducer;
import com.intellij.util.ui.ListTableModel;
import fr.adrienbrault.idea.symfony2plugin.Settings;
import fr.adrienbrault.idea.symfony2plugin.Symfony2ProjectComponent;
import fr.adrienbrault.idea.symfony2plugin.assistant.reference.AssistantReferenceUtil;
import fr.adrienbrault.idea.symfony2plugin.assistant.reference.MethodParameterSetting;
import fr.adrienbrault.idea.symfony2plugin.util.IdeHelper;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import javax.swing.event.TableModelEvent;
import javax.swing.event.TableModelListener;
import java.awt.*;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.util.ArrayList;
import java.util.List;

public class MethodParameterReferenceSettingsForm  implements Configurable {
    private JPanel panel1;
    private JPanel panelConfigTableView;
    private JButton buttonHelp;

    private TableView<MethodParameterSetting> tableView;
    private Project project;
    private boolean changed = false;
    private ListTableModel<MethodParameterSetting> modelList;

    public MethodParameterReferenceSettingsForm(Project project) {
        this.project = project;

        this.tableView = new TableView<MethodParameterSetting>();
        this.modelList = new ListTableModel<MethodParameterSetting>(
            new CallToColumn(),
            new MethodColumn(),
            new IndexColumn(),
            new ProviderColumn(),
            new ContributorColumn(),
            new ContributorDataColumn()
        );

        this.attachItems();

        this.tableView.setModelAndUpdateColumns(this.modelList);
        this.tableView.getModel().addTableModelListener(new TableModelListener() {
            @Override
            public void tableChanged(TableModelEvent e) {
                MethodParameterReferenceSettingsForm.this.changed = true;
            }
        });

        buttonHelp.addMouseListener(new MouseAdapter() {
            @Override
            public void mouseClicked(MouseEvent e) {
                super.mouseClicked(e);
                IdeHelper.openUrl(Symfony2ProjectComponent.HELP_URL + "extension/method_parameter.html");
            }
        });

    }

    private void attachItems() {
        for (MethodParameterSetting MethodParameterSetting : AssistantReferenceUtil.getMethodsParameterSettings(this.project)) {
            this.modelList.addRow(MethodParameterSetting);
        }
    }

    @Nls
    @Override
    public String getDisplayName() {
        return null;
    }

    @Nullable
    @Override
    public String getHelpTopic() {
        return null;
    }

    @Nullable
    @Override
    public JComponent createComponent() {
        ToolbarDecorator tablePanel = ToolbarDecorator.createDecorator(this.tableView, new ElementProducer<MethodParameterSetting>() {
            @Override
            public MethodParameterSetting createElement() {
                //IdeFocusManager.getInstance(TwigSettingsForm.this.project).requestFocus(TwigNamespaceDialog.getWindows(), true);
                return null;
            }

            @Override
            public boolean canCreateElement() {
                return true;
            }
        });

        tablePanel.setEditAction(new AnActionButtonRunnable() {
            @Override
            public void run(AnActionButton anActionButton) {
                MethodParameterReferenceSettingsForm.this.openTwigPathDialog(MethodParameterReferenceSettingsForm.this.tableView.getSelectedObject());
            }
        });


        tablePanel.setAddAction(new AnActionButtonRunnable() {
            @Override
            public void run(AnActionButton anActionButton) {
                MethodParameterReferenceSettingsForm.this.openTwigPathDialog(null);
            }
        });

        tablePanel.disableUpAction();
        tablePanel.disableDownAction();

        this.panelConfigTableView.add(tablePanel.createPanel());

        return this.panel1;
    }

    @Override
    public boolean isModified() {
        return this.changed;
    }

    @Override
    public void apply() throws ConfigurationException {
        List<MethodParameterSetting> methodParameterSettings = new ArrayList<MethodParameterSetting>();

        for(MethodParameterSetting methodParameterSetting :this.tableView.getListTableModel().getItems()) {
            methodParameterSettings.add(methodParameterSetting);
        }

        getSettings().methodParameterSettings = methodParameterSettings;
        this.changed = false;
    }

    private Settings getSettings() {
        return Settings.getInstance(this.project);
    }

    private void resetList() {
        // clear list, easier?
        while(this.modelList.getRowCount() > 0) {
            this.modelList.removeRow(0);
        }

    }

    @Override
    public void reset() {
        this.resetList();
        this.attachItems();
        this.changed = false;
    }

    @Override
    public void disposeUIResources() {

    }

    private class CallToColumn extends ColumnInfo<MethodParameterSetting, String> {

        public CallToColumn() {
            super("CallTo");
        }

        @Nullable
        @Override
        public String valueOf(MethodParameterSetting methodParameterSetting) {
            return methodParameterSetting.getCallTo();
        }
    }

    private class MethodColumn extends ColumnInfo<MethodParameterSetting, String> {

        public MethodColumn() {
            super("Method");
        }

        @Nullable
        @Override
        public String valueOf(MethodParameterSetting methodParameterSetting) {
            return methodParameterSetting.getMethodName();
        }
    }

    private class IndexColumn extends ColumnInfo<MethodParameterSetting, Integer> {

        public IndexColumn() {
            super("Index");
        }

        @Nullable
        @Override
        public Integer valueOf(MethodParameterSetting methodParameterSetting) {
            return methodParameterSetting.getIndexParameter();
        }
    }

    private class ProviderColumn extends ColumnInfo<MethodParameterSetting, String> {

        public ProviderColumn() {
            super("Provider");
        }

        @Nullable
        @Override
        public String valueOf(MethodParameterSetting methodParameterSetting) {
            return methodParameterSetting.getReferenceProviderName();
        }
    }

    private class ContributorColumn extends ColumnInfo<MethodParameterSetting, String> {

        public ContributorColumn() {
            super("Contributor");
        }

        @Nullable
        @Override
        public String valueOf(MethodParameterSetting methodParameterSetting) {
            return methodParameterSetting.getContributorName();
        }
    }

    private class ContributorDataColumn extends ColumnInfo<MethodParameterSetting, String> {

        public ContributorDataColumn() {
            super("ContributorData");
        }

        @Nullable
        @Override
        public String valueOf(MethodParameterSetting methodParameterSetting) {
            return methodParameterSetting.getContributorData();
        }
    }

    private void openTwigPathDialog(@Nullable MethodParameterSetting methodParameterSetting) {
        MethodParameterDialog twigNamespaceDialog;
        if(methodParameterSetting == null) {
            twigNamespaceDialog = new MethodParameterDialog(project, this.tableView);
        } else {
            twigNamespaceDialog = new MethodParameterDialog(project, this.tableView, methodParameterSetting);
        }

        Dimension dim = new Dimension();
        dim.setSize(500, 190);
        twigNamespaceDialog.setTitle("MethodParameterSetting");
        twigNamespaceDialog.setMinimumSize(dim);
        twigNamespaceDialog.pack();
        twigNamespaceDialog.setLocationRelativeTo(this.panel1);

        twigNamespaceDialog.setVisible(true);
    }

}

File: src/main/kotlin/fr/adrienbrault/idea/symfony2plugin/ui/MethodParameterReferenceSettingsForm.kt
package fr.adrienbrault.idea.symfony2plugin.ui

import com.intellij.openapi.options.Configurable
import com.intellij.openapi.project.Project
import com.intellij.ui.ToolbarDecorator
import com.intellij.ui.table.TableView
import com.intellij.util.ui.ColumnInfo
import com.intellij.util.ui.ElementProducer
import com.intellij.util.ui.ListTableModel
import fr.adrienbrault.idea.symfony2plugin.Settings
import fr.adrienbrault.idea.symfony2plugin.assistant.reference.AssistantReferenceUtil
import fr.adrienbrault.idea.symfony2plugin.assistant.reference.MethodParameterSetting
import fr.adrienbrault.idea.symfony2plugin.util.IdeHelper
import java.awt.BorderLayout
import java.awt.Dimension
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import javax.swing.JButton
import javax.swing.JLabel
import javax.swing.JPanel

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
class MethodParameterReferenceSettingsForm(private val project: Project) : Configurable {
    private val panel1 = JPanel(BorderLayout())
    private val panelConfigTableView = JPanel(BorderLayout())
    private val buttonHelp = JButton("Help")
    private val tableView = TableView<MethodParameterSetting>()
    private var changed = false
    private val modelList = ListTableModel<MethodParameterSetting>(
        CallToColumn(),
        MethodColumn(),
        IndexColumn(),
        ProviderColumn(),
        ContributorColumn(),
        ContributorDataColumn(),
    )

    init {
        createUIComponents()
        attachItems()

        tableView.setModelAndUpdateColumns(modelList)
        tableView.model.addTableModelListener { changed = true }

        buttonHelp.addMouseListener(object : MouseAdapter() {
            override fun mouseClicked(e: MouseEvent) {
                super.mouseClicked(e)
                IdeHelper.openUrl("https://www.jetbrains.com/help/phpstorm/symfony-creating-helper-functions.html#method-parameter")
            }
        })
    }

    private fun createUIComponents() {
        panel1.add(panelConfigTableView, BorderLayout.CENTER)

        val northPanel = JPanel(BorderLayout())
        northPanel.add(JLabel("Provides Completion and Goto for method parameter"), BorderLayout.CENTER)
        northPanel.add(buttonHelp, BorderLayout.EAST)
        panel1.add(northPanel, BorderLayout.NORTH)
    }

    private fun attachItems() {
        AssistantReferenceUtil.getMethodsParameterSettings(project).forEach(modelList::addRow)
    }

    override fun getDisplayName(): String? = null

    override fun getHelpTopic(): String? = null

    override fun createComponent(): JPanel {
        val tablePanel = ToolbarDecorator.createDecorator(tableView, object : ElementProducer<MethodParameterSetting> {
            override fun createElement(): MethodParameterSetting? {
                //IdeFocusManager.getInstance(TwigSettingsForm.this.project).requestFocus(TwigNamespaceDialog.getWindows(), true);
                return null
            }

            override fun canCreateElement(): Boolean {
                return true
            }
        })

        tablePanel.setEditAction { openTwigPathDialog(tableView.selectedObject) }
        tablePanel.setAddAction { openTwigPathDialog(null) }

        tablePanel.disableUpAction()
        tablePanel.disableDownAction()

        panelConfigTableView.add(tablePanel.createPanel())

        return panel1
    }

    override fun isModified(): Boolean = changed

    override fun apply() {
        settings.methodParameterSettings = ArrayList(tableView.listTableModel.items)
        changed = false
    }

    private val settings: Settings
        get() = Settings.getInstance(project)

    private fun resetList() {
        // clear list, easier?
        while (modelList.rowCount > 0) {
            modelList.removeRow(0)
        }
    }

    override fun reset() {
        resetList()
        attachItems()
        changed = false
    }

    private class CallToColumn : ColumnInfo<MethodParameterSetting, String>("CallTo") {
        override fun valueOf(item: MethodParameterSetting): String? = item.callTo
    }

    private class MethodColumn : ColumnInfo<MethodParameterSetting, String>("Method") {
        override fun valueOf(item: MethodParameterSetting): String? = item.methodName
    }

    private class IndexColumn : ColumnInfo<MethodParameterSetting, Int>("Index") {
        override fun valueOf(item: MethodParameterSetting): Int? = item.indexParameter
    }

    private class ProviderColumn : ColumnInfo<MethodParameterSetting, String>("Provider") {
        override fun valueOf(item: MethodParameterSetting): String? = item.referenceProviderName
    }

    private class ContributorColumn : ColumnInfo<MethodParameterSetting, String>("Contributor") {
        override fun valueOf(item: MethodParameterSetting): String? = item.contributorName
    }

    private class ContributorDataColumn : ColumnInfo<MethodParameterSetting, String>("ContributorData") {
        override fun valueOf(item: MethodParameterSetting): String? = item.contributorData
    }

    private fun openTwigPathDialog(methodParameterSetting: MethodParameterSetting?) {
        val dialog = if (methodParameterSetting == null) {
            MethodParameterDialog(project, tableView)
        } else {
            MethodParameterDialog(project, tableView, methodParameterSetting)
        }

        val dim = Dimension()
        dim.setSize(500, 190)
        dialog.title = "MethodParameterSetting"
        dialog.minimumSize = dim
        dialog.pack()
        dialog.setLocationRelativeTo(panel1)

        dialog.isVisible = true
    }
}
